pkg/controller/privatelink/conditions/conditions_test.go

```diff
@@ -0,0 +1,534 @@
+			}),
+		),
+		expectConditions: []hivev1.ClusterDeploymentCondition{{
+			Type:    hivev1.PrivateLinkFailedClusterDeploymentCondition,
+			Status:  corev1.ConditionFalse,
+			Reason:  testReason,
+			Message: testMessage,
+		}, {
+			Type:    hivev1.PrivateLinkReadyClusterDeploymentCondition,
+			Status:  corev1.ConditionFalse,
+			Reason:  testReason,
+			Message: testMessage,
+		}},
+	}, { // A false ready condition should be changed to true when completed is true
+		name: "false ready condition changed to true",
+		cd: cdBuilder.Build(
+			testcd.WithCondition(hivev1.ClusterDeploymentCondition{
+				Type:    hivev1.PrivateLinkFailedClusterDeploymentCondition,
+				Status:  corev1.ConditionFalse,
+				Reason:  testReason,
+				Message: testMessage,
+			}),
+			testcd.WithCondition(hivev1.ClusterDeploymentCondition{
+				Type:    hivev1.PrivateLinkReadyClusterDeploymentCondition,
+				Status:  corev1.ConditionFalse,
+				Reason:  testReason,
+				Message: testMessage,
+			}),
+		),
+		completed: corev1.ConditionTrue,
+		expectConditions: []hivev1.ClusterDeploymentCondition{{
+			Type:    hivev1.PrivateLinkFailedClusterDeploymentCondition,
+			Status:  corev1.ConditionFalse,
+			Reason:  testReason,
+			Message: testMessage,
+		}, {
+			Type:    hivev1.PrivateLinkReadyClusterDeploymentCondition,
+			Status:  corev1.ConditionTrue,
+			Reason:  testReason,
+			Message: testMessage,
+		}},
+	}, { // A true ready condition reason should be updated when completed is true
+		name: "true ready condition reason change",
+		cd: cdBuilder.Build(
+			testcd.WithCondition(hivev1.ClusterDeploymentCondition{
+				Type:    hivev1.PrivateLinkFailedClusterDeploymentCondition,
+				Status:  corev1.ConditionFalse,
+				Reason:  testReason,
+				Message: testMessage,
+			}),
+			testcd.WithCondition(hivev1.ClusterDeploymentCondition{
+				Type:    hivev1.PrivateLinkReadyClusterDeploymentCondition,
+				Status:  corev1.ConditionTrue,
+				Reason:  "some other reason",
+				Message: testMessage,
+			}),
+		),
+		completed: corev1.ConditionTrue,
+		expectConditions: []hivev1.ClusterDeploymentCondition{{
+			Type:    hivev1.PrivateLinkFailedClusterDeploymentCondition,
+			Status:  corev1.ConditionFalse,
+			Reason:  testReason,
+			Message: testMessage,
+		}, {
+			Type:    hivev1.PrivateLinkReadyClusterDeploymentCondition,
+			Status:  corev1.ConditionTrue,
+			Reason:  testReason,
+			Message: testMessage,
+		}},
+	}, { // There should be no changes when failed is already false and ready is already true with the same reason(s)
+		name: "no changes",
+		cd: cdBuilder.Build(
+			testcd.WithCondition(hivev1.ClusterDeploymentCondition{
+				Type:    hivev1.PrivateLinkFailedClusterDeploymentCondition,
+				Status:  corev1.ConditionFalse,
+				Reason:  testReason,
+				Message: testMessage,
+			}),
+			testcd.WithCondition(hivev1.ClusterDeploymentCondition{
+				Type:    hivev1.PrivateLinkReadyClusterDeploymentCondition,
+				Status:  corev1.ConditionFalse,
```

Comment:
This doesn't match the comment on L474.

I think if the ready condition is true but completed is false, we won't attempt to update the ready condition at all. Which would be inherently covered by changing this and L497 both to `ConditionTrue`. But we could prove that we don't even _attempt_ to update the ready condition in that path by making the pre-existing condition's reason and/or message different.

...and this behavior seems pretty odd to me. Why would we _not_ want to revert the ready condition from true to false if we call SetReadyCondition with `completed==ConditionFalse`? But I guess that's for discussion in the fup.

Reply:
Good catch... [fixed](https://github.com/openshift/hive/compare/651f14cd91272c1a380366df973b2f1365272c74..1adc65d49d7fa66ef436c8649fc3a53a67698d8b).

As for the ready condition, I'm not exactly sure. I think it limits flapping of the ready condition. When it is called after something is modified (ie, hzModified), then it always uses false. The only time it uses true is in the privatelink.go, and only after all of the actuator reconciles return true. I found that at least one of the awshubactuator functions always returned modified=true, so this would always flap false, and then back to true.

Once we move to the once-change-per-reconcile method (where I fixed this one function), it might make more sense to flap here. If you think this is still a desired change, we can continue the discussion in the followup.

Reply:
Okay, I figured it probably had something to do with preventing flapping. The part that doesn't make sense to me is how the Ready condition could have been set to True in the first place if we haven't actually completed the setup. That would seem to point to a bug somewhere in the code flow, and this odd logic was put in place to paper over it.

Alternatively the question may be: how did we become *un* ready after having become ready at some point? If that is legitimately possible, I would argue that we *should* be setting the Ready condition to false in that scenario.

But yeah, let's dig into it some more in the followup.

> Good catch... [fixed](https://github.com/openshift/hive/compare/651f14cd91272c1a380366df973b2f1365272c74..1adc65d49d7fa66ef436c8649fc3a53a67698d8b).

A thing I was trying to point out (albeit uneloquently) that this delta still doesn't cover: as written, we can't tell which path we took to end up no-op'ing: either the second branch of the weird anti-flapping condition, or the implicit `else` of that chunk. Since `completed` is false, we expect it to be the latter, but nothing about the test case proves that as currently written. My suggestion to remedy that was:

> But we could prove that we don't even _attempt_ to update the ready condition in that path by making the pre-existing condition's reason and/or message different.

That is, if the initial `Reason` was different, the `else if` path would revert it, but the implicit `else` path would not. The latter would be proven because we would have to change the `Reason` in the expected value as well.

Reply:
The flapping is happening after a successful reconcile (which sets the condition to true). In an ideal world, this -could- happen if one of the resources were deleted from the cloud provider. A future reconcile would identify that it does not exist and should report the resource as `modified`, causing a status (message) update.

However, in the code as currently written, [ReconcileHostedZoneRecords](https://github.com/openshift/hive/blob/master/pkg/controller/privatelink/actuator/awsactuator/awshubactuator.go#L371) always returns `modified` if there is no error. As a result, [recordsModified](https://github.com/openshift/hive/blob/20cdfee4b4d74d0281c6bd0f752a9de1113d3fef/pkg/controller/privatelink/actuator/awsactuator/awshubactuator.go#L154C5-L154C20) is always true and the status is always updated. Then it is updated again when the reconcile completes successfully.

I think a key point here is that `completed` != `ready`. Completed just indicates that we have completed the reconcile successfully and should update to a ready condition. In all other cases, we should just update the status message to match the latest change(s). At no point does not-completed indicate that ready condition should be false.

Reply:
Okay, so arguably both of those things are bugs -- which we can discuss in the fup.

For now all I'm asking is to change the `Reason` and/or `Message` to prove which code path is being taken.

Reply:
Doesn't the implicit `completed: False` mean it cannot logically be due to `else if completed == corev1.ConditionTrue {`?

I've [added](https://github.com/openshift/hive/compare/1adc65d49d7fa66ef436c8649fc3a53a67698d8b..5fc892e8809e53b5f2e7db379c45c5a54dc7565c) more `no change` cases to help test/clarify each of the three 1) if, 2) else-if, and 3) implicit else paths of the code.

Reply:
And [Done](https://github.com/openshift/hive/compare/5fc892e8809e53b5f2e7db379c45c5a54dc7565c..cad338665f4a84151936753250a71fade7add247).